pac4j-cas/src/main/java/org/pac4j/cas/credentials/extractor/TicketAndLogoutRequestExtractor.java

```diff
@@ -0,0 +1,170 @@
+package org.pac4j.cas.credentials.extractor;
+
+import org.jasig.cas.client.util.CommonUtils;
+import org.pac4j.cas.config.CasConfiguration;
+import org.pac4j.core.context.ContextHelper;
+import org.pac4j.core.context.HttpConstants;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.TokenCredentials;
+import org.pac4j.core.credentials.extractor.TokenCredentialsExtractor;
+import org.pac4j.core.exception.HttpAction;
+import org.pac4j.core.exception.TechnicalException;
+import org.pac4j.core.util.CommonHelper;
+import org.pac4j.core.util.InitializableWebObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.xml.bind.DatatypeConverter;
+import java.util.zip.Inflater;
+
+/**
+ * CAS ticket extractor or logout request handler.
+ *
+ * @author Jerome Leleu
+ * @since 1.9.2
+ */
+public class TicketAndLogoutRequestExtractor extends InitializableWebObject implements TokenCredentialsExtractor {
+
+    private final static int DECOMPRESSION_FACTOR = 10;
+
+    private final static Logger logger = LoggerFactory.getLogger(TicketAndLogoutRequestExtractor.class);
+
+    private CasConfiguration configuration;
+
+    private String clientName;
+
+    public TicketAndLogoutRequestExtractor() {}
+
+    public TicketAndLogoutRequestExtractor(final CasConfiguration configuration, final String clientName) {
+        this.configuration = configuration;
+        this.clientName = clientName;
+    }
+
+    @Override
+    protected void internalInit(final WebContext context) {
+        CommonHelper.assertNotNull("configuration", configuration);
+        CommonHelper.assertNotBlank("clientName", clientName);
+    }
+
+    @Override
+    public TokenCredentials extract(WebContext context) throws HttpAction {
+        // like the SingleSignOutFilter from the Apereo CAS client:
+        if (isTokenRequest(context)) {
+            final String ticket = context.getRequestParameter(CasConfiguration.TICKET_PARAMETER);
+            configuration.getLogoutHandler().recordSession(context, ticket);
+            final TokenCredentials casCredentials = new TokenCredentials(ticket, clientName);
+            logger.debug("casCredentials: {}", casCredentials);
+            return casCredentials;
+
+        } else if (isBackLogoutRequest(context)) {
+            final String logoutMessage = context.getRequestParameter(CasConfiguration.LOGOUT_REQUEST_PARAMETER);
+            logger.trace("Logout request:\n{}", logoutMessage);
+
+            final String ticket = CommonHelper.substringBetween(logoutMessage, CasConfiguration.SESSION_INDEX_TAG + ">", "</");
+            if (CommonUtils.isNotBlank(ticket)) {
+                configuration.getLogoutHandler().destroySessionBack(context, ticket);
+            }
+            final String message = "back logout request: no credential returned";
+            logger.debug(message);
+            throw HttpAction.ok(message, context);
+
+        } else if (isFrontLogoutRequest(context)) {
```

Comment:
You have some basic information about front channel logout here: https://apereo.github.io/cas/4.2.x/installation/Logout-Single-Signout.html

You're right: I don't think anyone uses it actually, but it certainly comes from the fragile design. I think it would make more sense to send multiple logout requests (let's say by hidden images on the logout page) with the message encrypted and base64 instead of relying on the multiple round-trips to perform the front channel logout. Maybe one of the last things remaining for CAS ;-)

That's why I developed the upgrade of the Apereo CAS client with the option of ignoring the RelayState and I did the same for pac4j.